lib/note-content-editor.jsx

```diff
@@ -63,6 +93,144 @@ export default class NoteContentEditor extends React.Component {
 		invoke( this, 'editor.focus' );
 	}
 
+	handleReturn = () => {
+		let editorState = this.state.editorState;
+		const key = editorState.getSelection().getFocusKey();
+		const block = editorState.getCurrentContent().getBlockForKey( key );
+		const line = block.getText();
+
+		// matches lines that start with `- ` or `* `
+		// preceded by 0 or more tab characters
+		const listItemRe = /^\t*[-\*]\s/;
+		let listItemMatch;
+
+		if ( line.trim() === '-' || line.trim() === '*' ) {
+			// if current line is just a bullet point and whitesapce,
+			// then clear the line (finishing a list)
+
+			// remove `- ` from the current line
+			editorState = EditorStateModifier.removeRange(
+				editorState,
+				SelectionState.createEmpty( key ).merge( {
+					focusOffset: block.getLength()
+				} )
+			)
+
+			// move selection to the start of the line
+			editorState = EditorState.forceSelection(
+				editorState,
+				SelectionState.createEmpty( key )
+			);
+
+			this.handleEditorStateChange( editorState );
+
+			return 'handled';
+		} else if ( ( listItemMatch = line.match( listItemRe ) ) ) {
+			// if the current line has a bullet point and text
+			// then insert a new line with a bullet point at the same indentation
+
+			// create a new line
+			editorState = EditorStateModifier.splitBlock( editorState );
+
+			// key of the new block
+			const nextKey = editorState.getCurrentContent().getKeyAfter( key );
+
+			// insert `- ` in the new line
+			editorState = EditorStateModifier.insertText(
+				editorState,
+				SelectionState.createEmpty( nextKey ),
+				listItemMatch[0]
+			);
+
+			// move selection to the end of the new line
+			editorState = EditorState.forceSelection(
+				editorState,
+				SelectionState.createEmpty( nextKey ).merge( {
+					anchorOffset: listItemMatch[0].length,
+					focusOffset: listItemMatch[0].length,
+				} )
+			);
+
+			// save new editorState
+			this.handleEditorStateChange( editorState );
+
+			return 'handled';
+		}
+	}
+
+	onTab = ( e ) => {
+		e.preventDefault()
+
+		let editorState = this.state.editorState;
+		const selection = editorState.getSelection();
+
+		if ( !selection.isCollapsed() ) {
+			return
+		}
+
+		const content = editorState.getCurrentContent();
+		const block = content.getBlockForKey( selection.getFocusKey() );
+		const selectionStart = selection.getStartOffset();
+
+		if ( !e.shiftKey ) {
+			const addTabAtStart = block.getText().trim() === '-'
+				|| block.getText().trim() === '*';
```

Comment:
Let's shorten this so that it fits comfortably on a single line or use some parenthesis or helper functions so that we're not stuck with an awkward split.

```js
const trimmedLine = block.getText().trim();
const addTabAtStart = trimmedText === '-' || trimmedText === '*';

const addTabAtStart = (
	trimmedText === '-' ||
	trimmedText === '*'
);

const addTabAtStart = [ '-', '*' ].contains( trimmedText );
```